### PR TITLE
Chant Create: Reimplement Suggested Chants feature

### LINKED ISSUE #1024: Chant Create suggested chant autofill buttons could be refactored to use a single function
Currently, on the Chant Create page, for the buttons in the sidebar providing suggested texts/genres for chants, we create a JavaScript function named after each suggestion's Cantus ID. Since Cantus IDs can contain `.`s and `:`s, and JS function names cannot, we have to convert Cantus IDs into valid alphanumeric strings for this to work.

A better possibility could be to create a single function into which we can pass a Cantus ID and it would handle all the autofilling. Discussion of how to implement this can be found in #1019.

Thanks, @ahankinson, for this suggestion!

### LINKED ISSUE #1388: Chant Create view: We need to reimplement the "suggested chants" feature
This feature was turned off by #1142, and we need to turn it back on again.

https://github.com/DDMAL/CantusDB/commit/92fa04590885aeb02a8475e547af5a6cf5110ad5 was the commit responsible for removing this functionality. (https://github.com/DDMAL/CantusDB/commit/fc188e94e2581fdb613e9948df1eec4309448196 and https://github.com/DDMAL/CantusDB/commit/d65f4246ab9e2870fec57a9660a3366d1cfd75b3 could also be useful for writing tests for this view)

I think it could be worth exploring implementing this feature client-side - when a user requests this page, we can return a page without sending a request to CI ourself. When the user's page loads, we get the user's browser to send requests to CI for the data necessary for creating this list of chants, and update the content of the page using JavaScript.
- It's less necessary to do this client-side than on publicly-accessible views, since we don't expect bots to be trawling through Chant Create pages.
- But this could allow us to return the page more quickly, allowing the user to start interacting with the page even before CI has returned its responses.

We can use the following endpoints:
- https://cantusindex.uwaterloo.ca/json-nextchants/{CantusID} to get the list of potential successor chants (note that CI's [documentation on this endpoint](https://github.com/dact-chant/cantus-index) indicates it is "under construction" - it seems to be working fine currently, but we should perhaps check in with Jan to confirm it will behave reliably)
- https://cantusindex.uwaterloo.ca/json-cid/{CantusID} to get the fulltext, genre, etc. of each potential successor chants

### PR DESCRIPTION
**Pull request by jacobdgm**
This PR reinstates the Suggested Chants buttons on the Chant Create page. It adds several functions in a new file, `cantusindex.py`. The main purpose of these functions is to request data from Cantus Index and parse it.

As part of writing these functions, I set up a system that allows us to replace `requests.get` calls with a mock function, so that CI doesn't get blitzed with requests whenever someone runs the test suite (and also so our test suite runs more quickly, and also so our tests don't start failing if the state of CI's database changes). These mock objects/functions are used extensively in `test_functions`, and are also used in `test_views.ChantCreateViewTest` wherever we had been requesting the Chant Create page. As we turn on features that request data from Cantus Index in other views, we should remember to patch any calls to `requests.get` in those views as well.

Once I had figured out how to refactor the `autoFillSuggestedChant` function(s) (see #1024), I went ahead and refactored the `autoFillFeast` functions as well.

This PR fixes #1388 and fixes #1024.

### CODE REVIEW

File: django/cantusdb_project/cantusindex.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        "fulltext": fulltext,
+        "incipit": incipit,
+        "genre_name": genre_name,
+        "genre_id": genre_id,
+        "clean_cantus_id": clean_cantus_id,
+    }
+
+
+def get_json_from_ci_api(
+    path: str, timeout: int = DEFAULT_TIMEOUT
```
Comment (dchiller) [inline django/cantusdb_project/cantusindex.py]:
```suggestion
    path: str, timeout: float = DEFAULT_TIMEOUT
```

`DEFAULT_TIMEOUT` is a float.

File: django/cantusdb_project/cantusindex.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        Optional[dict]: A dictionary with the following keys:
+            - "cantus_id"
+            - "occurrences"
+            - "fulltext"
+            - "incipit"
+            - "genre_id"
+            ...but if get_json_from_ci_api timed out, returns None instead
+    """
+    endpoint_path: str = f"/json-cid/{cantus_id}"
+    json: dict = get_json_from_ci_api(endpoint_path, timeout=timeout)
```
Comment (dchiller) [inline django/cantusdb_project/cantusindex.py]: Two things:
- This might be `None`. This case is handled below.
- This might be a List. Where is that handled?

File: django/cantusdb_project/cantusindex.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+# too high will cause the Chant Create page to take a very long time
+# to load. If/when we parallelize this process, we want to limit
+# the size of the burst of requests sent to CantusIndex.
+
+
+def get_suggested_chants(
+    cantus_id: str, number_of_suggestions: int = NUMBER_OF_SUGGESTED_CHANTS
+) -> Optional[list[dict]]:
+    endpoint_path: str = f"/json-nextchants/{cantus_id}"
+    all_suggestions: Optional[list] = get_json_from_ci_api(endpoint_path)
```
Comment (dchiller) [inline django/cantusdb_project/cantusindex.py]: Return type shows that this might be a `dict`. Is that handled?

File: django/cantusdb_project/cantusindex.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        # get_json_from_ci_api timed out
+        # or CI returned a response with no suggestions.
+        return None
+
+    # when Cantus ID doesn't exist within CI, CI's api returns a 200 response with `['Cantus ID is not valid']`
+    first_suggestion = all_suggestions[0]
+    if not isinstance(first_suggestion, dict):
+        return None
+
+    sort_by_occurrences: function = lambda suggestion: int(suggestion["count"])
```
Comment (dchiller) [inline django/cantusdb_project/cantusindex.py]: `function` is not a valid type. 
```suggestion
    sort_by_occurrences: Callable[[dict], int] = lambda suggestion: int(suggestion["count"])
```

File: django/cantusdb_project/cantusindex.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    sorted_suggestions: list = sorted(
+        all_suggestions, key=sort_by_occurrences, reverse=True
+    )
+    trimmed_suggestions: list = sorted_suggestions[:number_of_suggestions]
+
+    suggested_chants: list[dict] = []
```
Comment (dchiller) [inline django/cantusdb_project/cantusindex.py]:
```suggestion
    suggested_chants: list[Optional[dict]] = []
```

File: django/cantusdb_project/main_app/templates/chant_create.html [hunk 6]
```diff
@@ -255,11 +255,59 @@ <h3>Create Chant</h3>
             <div class="search-bar mb-3">
                 {% include "global_search_bar.html" %}
             </div>
-            <div class="card w-100">
+
+            <div class="card w-100 mb-3">
                 <div class="card-header">
-                    <h5><a id="source" href="{% url 'source-detail' source.id %}">{{ source.title }}</a></h5>
+                    <h5><a id="source" href="{% url 'source-detail' source.id %}">{{ source.siglum }}</a></h5>
```
Comment (dchiller) [inline django/cantusdb_project/main_app/templates/chant_create.html:245]: Is this intentional?

File: django/cantusdb_project/main_app/templates/chant_create.html [hunk 7]
```diff
@@ -255,11 +255,59 @@ <h3>Create Chant</h3>
...
+                                {% endif %}
+                            </div>
+                        {% if suggested_chants %}
+                            {% for suggestion in suggested_chants %}
+                                <input type="button" style="width: 80px" value="{{ suggestion.cantus_id }}" onclick="autoFill{{ suggestion.clean_cantus_id }}()" title="{{ suggestion.cantus_id }}"></input>
+                                <strong>{{ suggestion.genre_name }}</strong> - <span title="{{ suggestion.fulltext }}">{{ suggestion.incipit }}</span> (<strong>{{ suggestion.occurrences }}x</strong>)<br>
+                                <script>
+                                    function autoFill{{ suggestion.clean_cantus_id }}() {
+                                        document.getElementById('id_cantus_id').value = "{{ suggestion.cantus_id }}";
+                                        {% comment %} document.getElementById('id_genre').value = {{ suggestion.genre_id }}; {% endcomment %}
```
Comment (dchiller) [inline django/cantusdb_project/main_app/templates/chant_create.html]: Should this be deleted?

File: django/cantusdb_project/main_app/templates/chant_create.html [hunk 8]
```diff
@@ -255,11 +255,59 @@ <h3>Create Chant</h3>
...
+                                {% if previous_chant.cantus_id %}
+                                    (Cantus ID: <b><a href="https://cantusindex.org/id/{{ previous_chant.cantus_id }}" target="_blank">{{ previous_chant.cantus_id }}</a></b>)
+                                {% endif %}
+                            </div>
+                        {% if suggested_chants %}
+                            {% for suggestion in suggested_chants %}
+                                <input type="button" style="width: 80px" value="{{ suggestion.cantus_id }}" onclick="autoFill{{ suggestion.clean_cantus_id }}()" title="{{ suggestion.cantus_id }}"></input>
+                                <strong>{{ suggestion.genre_name }}</strong> - <span title="{{ suggestion.fulltext }}">{{ suggestion.incipit }}</span> (<strong>{{ suggestion.occurrences }}x</strong>)<br>
+                                <script>
+                                    function autoFill{{ suggestion.clean_cantus_id }}() {
```
Comment (dchiller) [inline django/cantusdb_project/main_app/templates/chant_create.html]: Why not have one function and then pass in arguments?

Comment (dchiller) [reply django/cantusdb_project/main_app/templates/chant_create.html]: E.g.: 

```javascript
function autoFill(cantus_id, full_text, genre){

}
```

File: django/cantusdb_project/main_app/templates/chant_create.html [hunk 9]
```diff
@@ -255,11 +255,59 @@ <h3>Create Chant</h3>
...
+                                        // in case suggestion.genre_id is None, don't try to change the #id_genre selector
+                                        if (genre_id == null) {
+                                            return
+                                        }
+                                        // Set the value, creating a new option if necessary
+                                        if ($('#id_genre').find("option[value='" + genre_id + "']").length) {
+                                            $('#id_genre').val(genre_id).trigger('change');
+                                        } else { 
+                                            // Create a DOM Option and pre-select by default
+                                            var newOption = new Option(genre_name, genre_id, true, true);
```
Comment (dchiller) [inline django/cantusdb_project/main_app/templates/chant_create.html]: Why is this necessary with genres? What is the situation where we want a genre that doesn't exist as an option to be created by a suggested chant?

File: django/cantusdb_project/main_app/tests/test_views.py [hunk 10]
```diff
@@ -2772,46 +2774,55 @@ def setUp(self):
...
-        response = self.client.post(
-            reverse("chant-create", args=[source.id]),
-            {
-                "manuscript_full_text_std_spelling": "initial",
-                "folio": "001r",
-                "c_sequence": "1",
-            },
-        )
+        with patch("requests.get", mock_requests_get):
+            response = self.client.post(
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_views.py]: Does a post request to `chant-create` make a call using `requests.get`?

File: django/cantusdb_project/main_app/tests/test_views.py [hunk 11]
```diff
@@ -2772,46 +2774,55 @@ def setUp(self):
...
+        with patch("requests.get", mock_requests_get):
+            response = self.client.get(url)
         self.assertEqual(response.context["source"].title, source.title)
 
     def test_post_error(self):
         """post with correct source and empty full-text"""
         source = make_fake_source()
         url = reverse("chant-create", args=[source.id])
-        response = self.client.post(url, data={"manuscript_full_text_std_spelling": ""})
+        with patch("requests.get", mock_requests_get):
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_views.py]: Same here re: post requests

File: django/cantusdb_project/main_app/tests/test_views.py [hunk 12]
```diff
@@ -2846,15 +2858,16 @@ def test_repeated_seq(self):
...
-        response = self.client.post(
-            url,
-            data={
-                "manuscript_full_text_std_spelling": fake_text,
-                "folio": TEST_FOLIO,
-                "c_sequence": random.randint(1, 4),
-            },
-            follow=True,
-        )
+        with patch("requests.get", mock_requests_get):
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_views.py]: Same here re: post request

File: django/cantusdb_project/main_app/tests/test_functions.py [hunk 13]
```diff
@@ -1,20 +1,162 @@
...
+            return MockResponse(
+                status_code=200,
+                content=mock_cantusindex_data.mock_json_cid_909030_content,
+                text=mock_cantusindex_data.mock_json_cid_909030_text,
+                json=mock_cantusindex_data.mock_json_cid_909030_json,
+            )
+        else:  # imitating CI's behavior when a made-up Cantus ID is entered.
+            return MockResponse(
+                status_code=500,
+                content=None,
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_functions.py:145]: `content` and `text` are not valid as `None` according to definition of `__init__` of `MockResponse`

File: django/cantusdb_project/main_app/tests/test_functions.py [hunk 14]
```diff
@@ -1,20 +1,162 @@
...
+    Args:
+        url (str): a URL - a necessary argument for requests.get
+        timeout (int): we pass timeout as an argument to requests.get in get_json_from_ci_api,
+            so mock_requests_get is configured to accept this argument.
+
+    Raises:
+        ValueError: This function is configured to mock requests to specific URLs only, including
+            - /json-nextchants/001010
+        If a call to requests.get with a different URL is made while mock_requests_get is patching it,
+        a ValueError is raised.
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_functions.py]:
```suggestion
        a NotImplementedError is raised.
```

File: django/cantusdb_project/main_app/tests/test_functions.py [hunk 15]
```diff
@@ -1,20 +1,162 @@
...
+
+    Returns:
+        MockResponse: A mock response object
+    """
+    if timeout < 0.001:
+        raise requests.exceptions.ConnectTimeout
+
+    if not (
+        "https://cantusindex.uwaterloo.ca" in url or "https://cantusindex.org" in url
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_functions.py]: Import this string from `cantusindex.py`?

File: django/cantusdb_project/main_app/tests/mock_cantusindex_data.py [hunk 16]
```diff
@@ -0,0 +1,644 @@
...
+            "db": "PEM"
+		}
+    }
+}
+"""
+mock_json_cid_909030_content: bytes = bytes(
+    mock_json_cid_909030_text,
+    encoding="utf-8-sig",
+)
+mock_json_cid_909030_json: dict = {
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/mock_cantusindex_data.py]: Is this the same as what is in `_content` and `_bytes`? Showing the conversion to json would make more sense to me.

File: django/cantusdb_project/main_app/tests/test_functions.py [hunk 17]
```diff
@@ -118,3 +260,156 @@ def test_generate_incipit(self):
...
+
+        # I can't figure out how to get assertRaises to work - even when the assertion should fail, it doesn't.
+        # It's not of vital importance that this argument check work correctly, I think, so I'm giving up in
+        # an effort not to get bogged down.
+        # Leaving a couple of commented-out attempts here with things I've tried, for the hopeful benefit of some
+        # future, cleverer developer
+        # - Jacob dGM, April 2024
+
+        # with self.subTest(test="Ensure raises ValueError when path improperly formatted, attempt 1"):
+        #     self.assertRaises(ValueError, get_json_from_ci_api, "/for/some/reason/this/passes/even/with/a/leading/slash")
```
Comment (dchiller) [inline django/cantusdb_project/main_app/tests/test_functions.py]: As I read it, your function `get_json_from_ci_api` would raise `ValueError` if there *isn't* a leading slash.

File: django/cantusdb_project/main_app/templates/chant_create.html [hunk 18]
```diff
@@ -282,9 +282,13 @@ <h5><a id="source" href="{% url 'source-detail' source.id %}">{{ source.siglum }
                                         document.getElementById('id_cantus_id').value = "{{ suggestion.cantus_id }}";
                                         {% comment %} document.getElementById('id_genre').value = {{ suggestion.genre_id }}; {% endcomment %}
                                         document.getElementById('id_manuscript_full_text_std_spelling').value = "{{ suggestion.fulltext }}";
-                                        var genre_id = {{ suggestion.genre_id }};
+                                        var genre_id = {{ suggestion.genre_id | default_if_none:"null" }};
                                         var genre_name = "{{ suggestion.genre_name }}";
 
+                                        // in case suggestion.genre_id is None, don't try to change the #id_genre selector
+                                        if (genre_id == null) {
```
Comment (ahankinson) [inline django/cantusdb_project/main_app/templates/chant_create.html]: Mmmmm. Python writing Jinja writing JavaScript. My favourite. 

You probably want `===` here.

File: django/cantusdb_project/main_app/templates/chant_create.html (hunk 6 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/templates/chant_create.html:245]: @annamorphism was looking over my shoulder as I did a bit of wireframing. This, iirc, is to make it match the sidebar cards on other pages.

But it should have been in its own commit. Thanks for checking on this.

@annamorphism, all this sounds right, yes?

File: django/cantusdb_project/main_app/templates/chant_create.html (hunk 18 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/templates/chant_create.html]: The question of `===` vs. `==` was on my mind while writing this. I think someone in a StackOverflow discussion said that `==` was correct, since professional JS developers will understand the difference between the two or some similar justification, but looking at it now, I'm inclined to agree with you. Let me make this change - thanks.

File: django/cantusdb_project/main_app/templates/chant_create.html (hunk 9 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/templates/chant_create.html]: This is necessary because we're using the django-autocomplete-light widget for the Genre selector. I had these same questions of @lucasmarchd01 when he implemented the Suggested Feasts feature - apparently, although it doesn't feel like it should be necessary to add options to the selector, this is the recommended way to use JS to change the options when using django-autocomplete-light. I copied and lightly adapted the code from later in this template.

See #1063.

File: django/cantusdb_project/cantusindex.py (hunk 3 above)
Comment (jacobdgm) [reply django/cantusdb_project/cantusindex.py]: You need to help me set up the Black VSCode extension sometime - thank you for pointing out all of the not-quite-correct type annotations in this PR...

File: django/cantusdb_project/main_app/templates/chant_create.html (hunk 9 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/templates/chant_create.html]: I've added comments to the places where we are setting both select elements.

File: django/cantusdb_project/main_app/tests/test_functions.py (hunk 17 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/tests/test_functions.py]: Right. So considering there _is_ a leading slash, and no `ValueError` is raised, we expect `AssertRaises` to fail. And yet it doesn't, for reasons I don't understand.

As far as I can figure, the most important part of a test suite is not that all the tests pass, but that the tests fail when they're supposed to. Before writing `self.assertRaises(ValueError, get_json_from_ci_api, "path/without/leading/slash")`, which passes, I want to be sure that `self.assertRaises(ValueError, get_json_from_ci_api, "/path/with/leading/slash")` fails.

File: django/cantusdb_project/main_app/tests/test_views.py (hunk 10 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/tests/test_views.py]: oh. this is a good point.

Let me think this through aloud: when we send a post request, the response will be an indication that the post request was received, along with a redirection status code indicating the next page the client should load. But `self.client.post` will not load the next page itself.

So long as my reasoning is correct, this patch is not necessary. Let me remove it. Thanks!

File: django/cantusdb_project/cantusindex.py (hunk 2 above)
Comment (jacobdgm) [reply django/cantusdb_project/cantusindex.py]: thank you

File: django/cantusdb_project/main_app/tests/test_functions.py (hunk 13 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/tests/test_functions.py:145]: thank you

File: django/cantusdb_project/main_app/templates/chant_create.html (hunk 6 above)
Comment (dchiller) [reply django/cantusdb_project/main_app/templates/chant_create.html:245]: Ok! That makes sense... I really was just making sure it didn't slip in unintentionally.

Comment (annamorphism) [reply django/cantusdb_project/main_app/templates/chant_create.html:245]: yes--OldCantus had the full title here (for some reason), which is probably why this got a title as well, but all the other similar sidebar cards use the siglum and we decided to make it match (better use of space, too.)

Comment (dchiller) [reply django/cantusdb_project/main_app/templates/chant_create.html:245]: Perfect! Going to "resolve"...

File: django/cantusdb_project/main_app/tests/test_functions.py (hunk 17 above)
Comment (dchiller) [reply django/cantusdb_project/main_app/tests/test_functions.py]: Ok, so this test does fail for me:

image

Comment (dchiller) [reply django/cantusdb_project/main_app/tests/test_functions.py]: It seems to me that if what I'm getting is the desired outcome, then this would be a place for an `expectedFailure` decorator?

Comment (jacobdgm) [reply django/cantusdb_project/main_app/tests/test_functions.py]: It's bizarre that this would fail for you but not for me.

I've also been having problems with one `ChantModelTest.test_get_concordances` - if I run `python manage.py test main_app.tests` as a whole, it errors, but if I run `python manage.py test main_app.tests.test_models.ChantModelTest`, it passes.

Did you try the `self.AssertRaises(several_arguments)` version or the `with self.AssertRaises(ValueError)` version? Should I just go ahead and uncomment that one and delete the other?

File: django/cantusdb_project/main_app/templates/chant_create.html (hunk 8 above)
Comment (jacobdgm) [reply django/cantusdb_project/main_app/templates/chant_create.html]: I think the bit I have been missing up until now is that I can potentially go `onclick="autoFillFeast({{feast.cantus_id}}, {{feast.name}}, etc.)`. Let me try to make this change. Thanks.

File: django/cantusdb_project/main_app/tests/test_functions.py (hunk 17 above)
Comment (dchiller) [reply django/cantusdb_project/main_app/tests/test_functions.py]: > I've also been having problems with one `ChantModelTest.test_get_concordances` - if I run `python manage.py test main_app.tests` as a whole, it errors, but if I run `python manage.py test main_app.tests.test_models.ChantModelTest`, it passes.

You should definitely make an issue about this...this is bad!

> Did you try the `self.AssertRaises(several_arguments)` version or the `with self.AssertRaises(ValueError)` version? Should I just go ahead and uncomment that one and delete the other?

The latter.

Comment (jacobdgm) [reply django/cantusdb_project/main_app/tests/test_functions.py]: oh! I think I figured it out.

This assertion was behaving weirdly because `"/path/with/leading/slash"` or whatever was not in the list of paths that `mock_requests_get` was set up to handle. Now, `mock_requests_get` raises a `NotImplementedError` when an unexpected URL/path is passed, but previously, it raised a - you guessed it - `ValueError`.

**Comment by jacobdgm** [comment]
I think I have addressed all the suggested changes! Thank you, @dchiller and @ahankinson, for your perceptive suggestions!

Once I figured out how to refactor the `autoFillSuggestedChants` functions, I went ahead and applied the same thing to the `autoFillFeast` functions.

**Review by dchiller — APPROVED** [review]
